lib/puma/minissl.rb

```diff
@@ -125,11 +125,14 @@ def flush
 
       def read_and_drop(timeout = 1)
         return :timeout unless IO.select([@socket], nil, nil, timeout)
-        return :eof unless read_nonblock(1024)
-        :drop
-      rescue Errno::EAGAIN
-        # do nothing
-        :eagain
+        case @socket.read_nonblock(1024, exception: false)
+        when nil
+          :eof
+        when :wait_readable
+          :eagain
+        else
+          :drop
+        end
```

Comment:
For the record, this might be related to a bug in OpenSSL, mentioned in https://github.com/ruby/openssl/issues/355#issuecomment-600875524
Anyway, it seems safer to not call to SSL anymore if we are in such a state, as this commit does.